File: dagger2-scopes-and-subcomponents-part3/app/src/main/java/com/example/dagger2scopesandsubcomponentspart3/AppModule.java
package com.example.dagger2scopesandsubcomponentspart3;


import dagger.Module;
import dagger.Provides;

@Module
public class AppModule {
    private int index;

    @Provides
    @ApplicationScope
    Warrior provideWarrior() {
        index++;
        return new Warrior("Warrior: " + index);
    }
}

File: dagger2-scopes-and-subcomponents/app/src/main/java/com/example/dagger2scopesandsubcomponents/AppModule.kt
package com.example.dagger2scopesandsubcomponents

import dagger.Module
import dagger.Provides

//@Module
//class AppModule {
//    private var index = 0
//
//    @Provides
//    fun provideWarrior(): Warrior {
//        index++
//        return Warrior("Warrior $index")
//    }
//}

@Module
class AppModule {
    private var index = 0

    @ApplicationScope
    @Provides
    fun provideWarrior(): Warrior {
        index++
        return Warrior("Warrior $index")
    }
}